data_safe_haven/pulumi/components/sre_dns_server.py

```diff
@@ -0,0 +1,319 @@
+"""Pulumi component for SRE DNS server"""
+import pathlib
+
+from pulumi import ComponentResource, Input, Output, ResourceOptions
+from pulumi_azure_native import containerinstance, network, resources
+
+from data_safe_haven.functions import (
+    allowed_dns_lookups,
+    b64encode,
+    bcrypt_encode,
+    ordered_private_dns_zones,
+)
+from data_safe_haven.pulumi.common import (
+    NetworkingPriorities,
+    SREDnsIpRanges,
+    SREIpRanges,
+    get_ip_address_from_container_group,
+)
+from data_safe_haven.utility import FileReader
+
+
+class SREDnsServerProps:
+    """Properties for SREDnsServerComponent"""
+
+    def __init__(
+        self,
+        admin_password: Input[str],
+        location: Input[str],
+        shm_fqdn: Input[str],
+        shm_networking_resource_group_name: Input[str],
+        sre_index: Input[int],
+    ) -> None:
+        subnet_ranges = Output.from_input(sre_index).apply(lambda idx: SREIpRanges(idx))
+        self.admin_password = Output.secret(admin_password)
+        self.admin_username = ("dshadmin",)
+        self.ip_range_prefix = str(SREDnsIpRanges().vnet)
+        self.location = location
+        self.shm_fqdn = shm_fqdn
+        self.shm_networking_resource_group_name = shm_networking_resource_group_name
+        self.sre_vnet_prefix = subnet_ranges.apply(lambda r: str(r.vnet))
+
+
+class SREDnsServerComponent(ComponentResource):
+    """Deploy DNS server with Pulumi"""
+
+    def __init__(
+        self,
+        name: str,
+        stack_name: str,
+        props: SREDnsServerProps,
+        opts: ResourceOptions | None = None,
+    ) -> None:
+        super().__init__("dsh:sre:DnsServerComponent", name, {}, opts)
+        child_opts = ResourceOptions.merge(opts, ResourceOptions(parent=self))
+
+        # Deploy resource group
+        resource_group = resources.ResourceGroup(
+            f"{self._name}_resource_group",
+            location=props.location,
+            resource_group_name=f"{stack_name}-rg-dns",
+            opts=child_opts,
+        )
+
+        # Read AdGuardHome setup files
+        resources_path = (
+            pathlib.Path(__file__).parent.parent.parent / "resources" / "dns_server"
+        )
+        adguard_entrypoint_sh_reader = FileReader(resources_path / "entrypoint.sh")
+        adguard_adguardhome_yaml_reader = FileReader(
+            resources_path / "AdGuardHome.mustache.yaml"
+        )
+
+        # Expand AdGuardHome YAML configuration
+        adguard_adguardhome_yaml_contents = Output.all(
+            admin_username=props.admin_username,
+            admin_password_encrypted=props.admin_password.apply(
+                lambda p: bcrypt_encode(p)
+            ),
+            upstream_dns="168.63.129.16",
+            filter_deny=["*.*"],
```

Comment:
Could the filter just be `*`? I suppose everything has a TLD so `*.*` always matches too?

Reply:
I found that `*` didn't work for some reason. Other "block-everything" lists use `*.*` too. I agree that either should be fine though.